tests/migration/postcopy.go

```diff
@@ -0,0 +1,399 @@
+								return nil
+							}
+							return fmt.Errorf("waiting for virt-handler pod to come back online")
+						}, 120*time.Second, 1*time.Second).Should(Succeed(), "Virt handler should come online")
+					}
+					It("should make sure that VM restarts after failure", func() {
+
+						By("creating a large VM with RunStrategyRerunOnFailure")
+						vm := createLargeVirtualMachine(testsuite.NamespacePrivileged)
+
+						// update the migration policy to ensure slow pre-copy migration progress instead of an immidiate cancelation.
+						migrationPolicy.Spec.CompletionTimeoutPerGiB = kvpointer.P(int64(20))
+						migrationPolicy.Spec.BandwidthPerMigration = kvpointer.P(resource.MustParse("1Mi"))
+						applyKubevirtCR()
+
+						By("Starting the VirtualMachine")
+						vm = tests.RunVMAndExpectLaunchWithRunStrategy(virtClient, vm, v1.RunStrategyRerunOnFailure)
+						vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &metav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
+
+						By("Checking that the VirtualMachineInstance console has expected output")
+						Expect(console.LoginToFedora(vmi)).To(Succeed())
+
+						// Need to wait for cloud init to finish and start the agent inside the vmi.
+						Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
+
+						runStressTest(vmi, "350M", stressDefaultSleepDuration)
+
+						By("Starting the Migration")
+						migration := libmigration.New(vmi.Name, vmi.Namespace)
+						migration = libmigration.RunMigration(virtClient, migration)
+
+						// check VMI, confirm migration state
+						waitUntilMigrationMode(vmi, v1.MigrationPostCopy, 300)
+
+						// launch killer pod on every node that isn't the vmi's node
+						By("Starting migration killer pods")
+						runMigrationKillerPod(vmi.Status.NodeName)
+
+						By("Making sure that post-copy migration failed")
+                        Eventually(matcher.ThisMigration(migration), 150, 1*time.Second).Should(BeInPhase(v1.MigrationFailed))
+
+						By("Removing migration killer pods")
+						removeMigrationKillerPod()
+
+						By("Ensuring the VirtualMachineInstance is restarted")
+						Eventually(ThisVMI(vmi), 240*time.Second, 1*time.Second).Should(beRestarted(vmi.UID))
+					})
+				})
+			})
+		})
+	})
+})
+
+func beRestarted(oldUID types.UID) gomegatypes.GomegaMatcher {
+	return gstruct.PointTo(gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+		"ObjectMeta": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+			"UID": Not(Equal(oldUID)),
+		}),
+		"Status": gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
+			"Phase": Equal(v1.Running),
+		}),
+	}))
+}
+
+func VMIMigrationWithGuestaAgent(virtClient kubecli.KubevirtClient, pvName string, memoryRequestSize string, migrationPolicy *migrationsv1.MigrationPolicy) {
```

Comment:
nit, typo:
`VMIMigrationWithGuestAgent`
IMO we should do something to this "test" shared between two files. But definitely outside of the scope of this PR.
So we should do it in a follow up :)